Store forum token instead of password

RoiEXLab opened the pull request:
<!--
  Commit comment above summarizing the update.  If multiple commits please
  summarize the change above. Commit comments should include an overview of
  the updates and the goal and reasoning behind the update.
  Code standards and PR guidelines can be found at:
  - https://github.com/triplea-game/triplea/wiki/Contribution-Guidelines
  - https://github.com/triplea-game/triplea/wiki/Code-Reviews
-->

This is what I was able to come up with from #6452
This PR is a WIP, but I wanted to hear your thoughts on the UI decisions.

Basically the user never really gets to see the token and we only store the username for display purposes, so the user can tell their settings have some sort of credentials stored.
We also store a uid to be able to properly revoke the token again.

Whenever a user clicks "save" now in engine preferences, we make a request to the proper forum and create a new token. There's no UI mechanism in NodeBB for this outside of the admin panel. Tokens are revoked if
1. A User creates a new token aka changes their username in the UI and clicks save
2. The Setting is cleared.

This has 2 benefits:
1. It works with 2FA which is the main goal of this PR
2. We no longer have to store any forum password, just a random token which is huge for security (maybe we want to consider explicitly clearing stored passwords to prevent them from floating around in the registry?)

Open question: How should we implement this on the "on screen"/not-engine-preferences login fields? Same UI like the engine preferences or just a pre-filled token field?

## Functional Changes
<!-- Put an X next any that apply -->
[] New map or map update
[] New Feature
[x] Feature update or enhancement
[] Feature Removal
[] Code Cleanup or refactor
[] Configuration Change
[] Problem fix:  <!-- Link to bug issue or forum post here -->
[] Other:   <!-- Please specify -->

## Testing
<!--
  Describe any manual testing performed below.
-->

I verified basic posting works with this new approach.

<!-- If there are UI updates, uncomment and include screenshots below -->
<!--
## Screens Shots

### Before

### After
-->

<!--
  Uncomment the below and add any additional details that would be helpful for reviewers.
-->
<!--
## Additional Review Notes
-->

## Release Note

<!--
Include a release note if there is a bug fix or a visible change for players.
For format & syntax help, see:
https://github.com/triplea-game/triplea/wiki/PR-Release-Notes
-->

<!--RELEASE_NOTE-->Feature|Support for 2FA for PbF<!--END_RELEASE_NOTE--><br>
<!--RELEASE_NOTE-->Enhancement|PbF now stores login tokens instead of passwords<!--END_RELEASE_NOTE--><br>
<!--RELEASE_NOTE-->Fix|Usernames with spaces work for PbF<!--END_RELEASE_NOTE-->

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java:
```diff
@@ -120,7 +120,10 @@ synchronized void setTopicId(final String topicId) {
+    return isTopicIdValid()
+        && ((isForumUsernameValid() && isForumPasswordValid())
+            || ClientSetting.aaForumToken.isSet()
+            || ClientSetting.tripleaForumToken.isSet());
```
This class is just band-aided so I could use the "Test PbF" button and everything compiles properly.
This will obviously change in the final verison of this PR, but I haven't had them time for everything today.

RoiEXLab commented on game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java:
```diff
@@ -724,14 +727,28 @@ public void reset() {
-          new JPasswordField(credentialToString(passwordSetting::getValue), 20);
+          new JTextFieldBuilder()
+              .columns(20)
+              .text(usernameSetting.getValue().map(String::new).orElse(""))
```
I will properly clear any char arrays in the final form as well, even though I'm not sure if actually encrypting everything is that important now

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java:
```diff
@@ -88,10 +88,10 @@ synchronized void setForumSelection(final String forumSelection) {
             ? ClientSetting.tripleaForumUsername.getValue().map(String::valueOf).orElse("")
             : ClientSetting.aaForumUsername.getValue().map(String::valueOf).orElse("");
-    forumPasswordIsSet =
+    forumPasswordIsSet = false; /*
```
If I understand correctly, we'll use 'username' + 'password' to create a token via API, then store that token but not the password (correct?).

If so, I think we probably want this field to check if the token is set. If the token is set, then we can fill in the password field with a dummy value to make it look like there is something filled in (that should be essentially what we have now - the reason for that is so we don't have to read the password from settings and set the actual value to the text field. Instead we use a dummy string to show that the field is set).

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,129 @@
+  private void deleteToken(final CloseableHttpClient client, final int userId, final String token)
+      throws IOException {
+    final HttpDelete httpDelete =
+        new HttpDelete(forumUrl + "/api/v2/users/" + userId + "/tokens/" + token);
```
Any thoughts to first, or as part of this PR, to convert the http client to Feign?

I mention that for a few reasons:
- the different styles/technologies for http communication is inconsistent. It's one of the technology consistencies that a project really should try to keep so you limit how many technologies you need to learn and how many different patterns there are for doing something.
- the test infrastructure for http communication with feign client is pretty straight forward. At this point with wiremock, we should be testing this client, going to a consistent pattern and using the same technological infrastructure would make that easy.

RoiEXLab replied:
I was thinking about migrating away from http component to either use the built-in java API that's pretty neat or feign, but not as part of this PR

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,129 @@
+        new HttpDelete(forumUrl + "/api/v2/users/" + userId + "/tokens/" + token);
+    HttpProxy.addProxy(httpDelete);
+    httpDelete.addHeader("Authorization", "Bearer " + token);
+    client.execute(httpDelete).close(); // ignore errors, execute and then close
```
ignore error is a bit scary :grin: 

Should we at least info log them? Have you seen this throw errors?

RoiEXLab replied:
This is super funny, because you introduced this change in the first place ^^

Took me a while to find the commit though: https://github.com/triplea-game/triplea/pull/3715/files#diff-317b30fefdd065ba13162cb51844ab22L93-R96

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,129 @@
+
+  private int getUserId(final CloseableHttpClient client, final String username)
+      throws IOException {
+    final Map<?, ?> jsonObject = queryUserInfo(client, username);
```
Bit nitpicky, I think we should try to avoid `jsonObject`. The debian variant of TripleA ripped these out, the lack of type-safety is not great. Going to Feign makes this mapping to Java object automatic. If migrating to Feign, this would be moot, otherwise have you considered using GSON?

RoiEXLab replied:
This used to be a json dependency, but it is currently just using the snakeyaml-engine dependency wich supports the yaml 1.2 standard -> json support.

Feign would be an option for a follow-up PR, but out of scope for this one

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,129 @@
+      throw new IllegalStateException(String.format("User %s doesn't exist.", username));
+    }
+    if (1 == (Integer) jsonObject.get("banned")) {
+      throw new IllegalStateException("Your account is banned from the forum.");
```
nit, we probably want to take care about exception types. Exceptions that we may show to users we would want to be probably first-class exceptions. Maybe even more ideally not exceptions at all. For example this checkUser method could return a status flag, that would eliminate control-flow-by-exception. What do you think?

RoiEXLab replied:
I agree the Exceptions are not ideal, but to get this done I just copied the implementation from the forum poster class, so I didn't really change anything here

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java:
```diff
@@ -749,27 +768,67 @@ public JComponent getUiComponent() {
+                  final TokenInfo tokenInfo =
+                      tokenGenerator.generateToken(
+                          usernameField.getText(),
+                          new String(passwordField.getPassword()),
```
I know we had the discussion about passwords as `String` being security theater. IMO if we can keep passwords as `char[]` without too much trouble, it's not a bad thing. Said another way, if we don't have to go too far out of our way to use `char[]` for password, we probably should.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java:
```diff
@@ -749,27 +768,67 @@ public JComponent getUiComponent() {
+          return;
+        }
+        try {
+          BackgroundTaskRunner.runInBackground(
```
Side-note, we really need to update BackTasRunner to be closeable. If anything crashes  while this is running we can run into problems. Notably if any exception is uncaught, then is handled and shows the error dialog, the background task will not complete until the dialog is closed and the dialog cannot be closed because the background task runner is modal and does not give up UI focus.

We'll probably want to test this out and/or fix the background task runner. We could see problems if the lobby were not responding or is being restarted, in such a case we'd want a graceful failure.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java:
```diff
@@ -749,27 +768,67 @@ public JComponent getUiComponent() {
+                final NodeBbTokenGenerator tokenGenerator = new NodeBbTokenGenerator(forumUrl);
+                final Optional<Integer> oldUserId = uidSetting.getValue();
+                final Optional<char[]> oldToken = tokenSetting.getValue();
+                if (!usernameField.getText().isEmpty()) {
```
Perhaps we want to use `isBlank()` instead of `isEmpty()`.

Do we currently already enforce a minimum length check on usename before the 'save' button is enabled? It might be the case that we've already guaranteed that there is a non-empty value that is of a minimum length before we hit this code which would make this validation redundant.

Side-comment, redundant client-side validation can generally be removed. If someone hacks it up so a button is enabled without validation, they can hack this. Server validation needs to be redundant in case a client hacks out the validation (or is just using straight-up `curl` or something similar).

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java:
```diff
@@ -749,27 +768,67 @@ public JComponent getUiComponent() {
       public void reset() {
-        usernameField.setText(credentialToString(usernameSetting::getValue));
-        passwordField.setText(credentialToString(passwordSetting::getValue));
+        usernameField.setText(usernameSetting.getValue().map(String::new).orElse(""));
```
I don't think `username` should be a `char[]`, IMO it's just not sensitive enough. Does it make sense changing `username` to be a String value before, after or in this PR?

RoiEXLab replied:
Yes and no.
1. It would make sense changing it to a normal string, but for usernames that are actually stored already it would fill the username field with a cryptic string for the first time, so we'd have to change the name of the setting to something different.

Regardless I think we should offer a button that performs a "factory reset" on the preferences store so users have a way to clear out settings that might no longer be there at all.

DanVanAtta replied:
IF we change datatype of a client setting, we only need to change the name of the setting. Having something clear out old settings would be nice, but I think certainly nice and not required.

RoiEXLab commented:
> Off-topic, this makes me wonder if we can do the same for email. Handling any form of email password just really worries me.

For certain services that offern an API like google, the answer is probably yes, but in general the answer is probably no, because every email provider handles things differently

DanVanAtta commented:
@RoiEXLab any thoughts when you might be able to wrap this up?

RoiEXLab commented:
Coding can probably be done in a matter of half an hour or something.
My hesitation comes partly from my weird weekly schedule partly because I'm still not sure how to "redesign" the UI elements inside the PbF Menu so they work with the new system.
Pre-filling a password doesn't work anymore, so how do we indicate a token has already been provided? Do we  simply disable the forum and add a tooltip? Do we fill in some stars to indicate _some_ password, just to confuse people that they entered a password with a different amount of characters?
Apart from that I'm not that happy with the temporary storage of the currently password, then token.
Resetting the value requires us to "unregister" the token, which might not be an ideal thing to do in a shutdown hook.

If you think you have a good solution for all of those questions I'd be happy to implement them on the weekend or something, but I don't want to spend 2 hours just trying things until I get to a point that somehow works

DanVanAtta commented:
IMO setting a long 'dummy' password value is reasonable (ballpark 15~20 characters). I think in the case of saved passwords we are already pre-filling a dummy value in the password field to show that it has been set (so that is not without precedent).

> Resetting the value requires us to "unregister" the token

I'm not sure I have enough context to really suggest a solution to this. The 'unregister' is with client settings, or with the forum?

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,129 @@
+    try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
+      deleteToken(client, userId, token);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to revoke login token", e);
```
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbForumPoster.java:
```diff
@@ -66,10 +59,7 @@
-        Arrays.withSensitiveArrayAndReturn(() -> forumPostingParameters.username, String::new);
-    this.password =
-        Arrays.withSensitiveArrayAndReturn(() -> forumPostingParameters.password, String::new);
+    this.token = new String(forumPostingParameters.token);
```
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,129 @@
+      final int userId = getUserId(client, username);
+      return new TokenInfo(getToken(client, userId, password, otp), userId);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to retrieve login token", e);
```
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,131 @@
+    try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
+      deleteToken(client, userId, token);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to revoke login token", e);
```
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,131 @@
+      final int userId = getUserId(client, username);
+      return new TokenInfo(getToken(client, userId, password, otp), userId);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to retrieve login token", e);
```
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,131 @@
+
+  private Map<?, ?> queryUserInfo(final CloseableHttpClient client, final String username)
+      throws IOException {
+    final String encodedUsername = URLEncoder.encode(username, StandardCharsets.UTF_8);
```
Note this fix for #6737 I added

DanVanAtta replied:
@RoiEXLab we should probably put the bug fix to its own PR:
- when this gets squashed, a bug fix being mixed in with a feature update will be somewhat confusing
- the bugfix needs to go out with the next release, which could be as soon as tonight or tomorrow.

DanVanAtta replied:
Would be nice to see a test case for this as well considering it was a bug. Tests are good for preventing future regressions, ensuring bugs are fixed well, and for giving feedback if the code is designed well, which should drive generally better code design (or at least prevent completely ad-hoc and  chaotic design).

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java:
```diff
@@ -41,13 +41,15 @@
   private final JTextField usernameField = new JTextField();
   private final JLabel passwordLabel = new JLabel("Forum Password");
   private final JPasswordField passwordField = new JPasswordField();
+  private final JLabel otpLabel = new JLabel("2FA OTP-Code (optional)");
```
`otp` is not that well known of an acronym. What is it spelled out?

Can we add a help text or box next to this as well to explain what this is for a user? We *do* want people to use this after all. If we look at the 1.9 PBF menu, we can see that the game explained pretty much all the options quite well. This meant any user could learn how to use PBF, even if it was kinda clunky.

2FA OTP-code is a bit cryptic to even me. I think we'll need to help users a bit more, and/or explain when/why they'd want to use this. I'm a bit curious myself when we would want to use OTP code and not just use passwords.

RoiEXLab replied:
OTP is short for "One-Time-Password".

This Field is only required for people that have 2FA enabled in the Forum. Up until recently the "write-api" for NodeBB didn't check if 2FA was enabled and just ignored it. Because @panther2 pointed this out I created an issue which lead to a fix. Currently people with 2FA enabled can't use PbF at all and this extra field should fix this

RoiEXLab replied:
Also perhaps worth mentioning: The OTP isn't a replacement for the password, it is a necessary addition

DanVanAtta replied:
Cool, that is good context. Do you have a plan for how to convey that information to users? Specifically, what the field is, when it is needed, and if then not already clear - perhaps where to get their OTP when it is needed.

RoiEXLab replied:
Not really, but I think this is a problem that isn't limited to TripleA. How can we explain or even advertise 2FA for users?
For most people this just adds another inconvenient step to the login process, and I'd assume most people already using 2FA know what to look out for, they don't need any extra explanation, they just want a field to enter their newly generated code.

If you have any good explanation, please let me know. I'd also like to know how to explain such an "abstract" concept in such a limited screen space 🤔

RoiEXLab replied:
To clarify: Obviously we could implement a wall of text, but I bet most people aren't going to read it if it's more than a couple of lines

DanVanAtta replied:
A help button that says this field is used if "2FA is enabled. That this value can be found in forum settings. Doesn't have to be anything to complicated. Leaving it as 'OTP' though is going to be cryptic.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java:
```diff
@@ -63,6 +65,10 @@
     this.viewModel = viewModel;
     viewModel.setView(this);
+    // If password is already stored we already remember
+    if (viewModel.isForumPasswordValid()) {
```
nit, view should not be doing logic if it has a view model. The view model should be the one that owns the full logic of when 'rememberPassword' is selected.

Would recommend to fix by moving the logic to view model, EG:
```
rememberPassword.setSelected(viewModel.isRememberPasswordSelected());
```

RoiEXLab replied:
To be honest this is more like a hack.
If the user already has a token stored, it makes no sense to have "Remember Password" unchecked, which would cause the credentials to be removed on close by default.
On the other hand if no token is stored it "defaults" to the ClientSetting that stores the last state. That's why the code is the way it is.

I think "Remember Password" shouldn't be a client setting at all. The checkbox should be selected by default if and only if the token is already stored and leave it up to the user to alter this behaviour.

DanVanAtta replied:
The 'remember password' is a setting as we are currently supporting passwords, and if you select the box then you want it checked, and if you unselect it then it'll be annoying to have to unselect it every time. Those users are probably very worried about their password and if they forget to unselect the value then they'll be unhappy with the software for allowing the error.

It sounds like if 'remember password' does not apply with tokens then it should be hidden.

Either way, any `if` statements should be removed from the view and moved to the model. The view should not know how to compute the UI state, it should ask the model what its appropriate state should be and render that.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,131 @@
+import org.snakeyaml.engine.v2.api.LoadSettings;
+
+@AllArgsConstructor
+public class NodeBbTokenGenerator {
```
Could use a javadoc on this class IMO to explain it's general usage and how it works.

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,155 @@
+      final int userId = getUserId(client, username);
+      return new TokenInfo(getToken(client, userId, password, otp), userId);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to retrieve login token", e);
```
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java:
```diff
@@ -0,0 +1,155 @@
+    try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
+      deleteToken(client, userId, token);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to revoke login token", e);
```
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

DanVanAtta replied:
Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

RoiEXLab commented:
@DanVanAtta I added documentation and updated the PR description. Let me know what you think so far. I only had a brief look at #6925 so far but I think most of the few remaining open questions can be dicussed there

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditorViewModel.java:
```diff
@@ -7,19 +7,27 @@
 import org.triplea.java.ViewModelListener;
 
 class ForumPosterEditorViewModel {
+  private static final int DUMMY_PASSWORD_LENGTH = 4;
```
Oh.. the UX should be changed then probably. Perhaps a button to log in to the forum, clicking that asks for username and password, and once done then the user sees a "credentials saved" label somewhere and are given a button to clear their credentials. When entering in their credentials, that would be a good place to have a checkbox for "clear credentials when the application closes". 

This way we never need to enter in a dummy password at all, we don't have to worry about a remember password checkbox, and if the credentials are already entered we don't show needless text fields on the UI.

DanVanAtta commented:
@RoiEXLab I want the URI decode bug fix for 2.1, sooner. To get there we *should* create a test for it, but to test it we should be extracting the HTTP client portion so that we can get at the logic that needs to be tested. I am thinking we probably should first convert the HTTP client to feign so it's uniform and testable, then do this work.

On another front, I also think the completable future portions can be removed as we are simply calling `get` on the future and blocking anyways. I think then the error handling is perhaps not the best as we catch `IllegalStateException` (a smell) and use the message from the exception as an error message to user. Since you can trigger the exception by simply using the system in a very normal way, it's control-flow-by-exception handling. A user should never be able to trigger an exception via normal usage and standard clicks (pulling the network cable is questionable and probably an okay exception, but simply not completing registration or typo in username are not exceptions, those are normal, they are "user-error" cases, not exceptions).

So, if we convert the return value from the http client to be a result and result code, we can remove the completable futures and an exception would be an IOException only (which we would perhaps want reported and/or retried).  I digress though, one main point is if we don't convert to Feign now and test now, I don't think we'll do either one ever.

DanVanAtta commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java  7
- game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbf/ForumPosterEditor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573&bid=18410695)

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbForumPoster.java:
```diff
@@ -67,10 +59,7 @@
-        Arrays.withSensitiveArrayAndReturn(() -> forumPostingParameters.username, String::new);
-    this.password =
-        Arrays.withSensitiveArrayAndReturn(() -> forumPostingParameters.password, String::new);
+    this.token = new String(forumPostingParameters.token);
```
Codacy found an issue: [Avoid instantiating String objects; this is usually unnecessary.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5619573)

RoiEXLab commented:
Going to merge this now. Explanation can be found in #6925 

Ultimate goal is to migrate this to feign to allow for less logic and nicer tests
